api/proto/teleport/okta/v1/okta_service.proto

```diff
@@ -0,0 +1,141 @@
+// Copyright 2023 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+package teleport.okta.v1;
+
+import "google/protobuf/empty.proto";
+import "teleport/legacy/types/types.proto";
+
+option go_package = "github.com/gravitational/teleport/api/gen/proto/go/teleport/okta/v1";
+
+// OktaService provides CRUD methods for Okta resources.
+service OktaService {
+  // ListOktaImportRules returns a paginated list of all Okta import rule resources.
+  rpc ListOktaImportRules(ListOktaImportRulesRequest) returns (ListOktaImportRulesResponse);
+  // GetOktaImportRule returns the specified Okta import rule resources.
+  rpc GetOktaImportRule(GetOktaImportRuleRequest) returns (types.OktaImportRuleV1);
+  // CreateOktaImportRule creates a new Okta import rule resource.
+  rpc CreateOktaImportRule(CreateOktaImportRuleRequest) returns (types.OktaImportRuleV1);
+  // UpdateOktaImportRule updates an existing Okta import rule resource.
+  rpc UpdateOktaImportRule(UpdateOktaImportRuleRequest) returns (types.OktaImportRuleV1);
+  // DeleteOktaImportRule removes the specified Okta import rule resource.
+  rpc DeleteOktaImportRule(DeleteOktaImportRuleRequest) returns (google.protobuf.Empty);
+  // DeleteAllOktaImportRules removes all Okta import rules.
+  rpc DeleteAllOktaImportRules(DeleteAllOktaImportRulesRequest) returns (google.protobuf.Empty);
+
+  // ListOktaAssignments returns a paginated list of all Okta assignment resources.
+  rpc ListOktaAssignments(ListOktaAssignmentsRequest) returns (ListOktaAssignmentsResponse);
+  // GetOktaAssignment returns the specified Okta assignment resources.
+  rpc GetOktaAssignment(GetOktaAssignmentRequest) returns (types.OktaAssignmentV1);
+  // CreateOktaAssignment creates a new Okta assignment resource.
+  rpc CreateOktaAssignment(CreateOktaAssignmentRequest) returns (types.OktaAssignmentV1);
+  // UpdateOktaAssignment updates an existing Okta assignment resource.
+  rpc UpdateOktaAssignment(UpdateOktaAssignmentRequest) returns (types.OktaAssignmentV1);
+  // DeleteOktaAssignment removes the specified Okta assignment resource.
+  rpc DeleteOktaAssignment(DeleteOktaAssignmentRequest) returns (google.protobuf.Empty);
+  // DeleteAllOktaAssignments removes all Okta assignments.
+  rpc DeleteAllOktaAssignments(DeleteAllOktaAssignmentsRequest) returns (google.protobuf.Empty);
```

Comment:
nit: definitely a personal preference thing, but I usually try to fold similar methods into one another.  E.g. `DeleteAll` isn't necessary if the resource ID in `Delete` is optional, `Get` is unnecessary if `List` includes an optional resource ID field, etc.

I don't think there's really a right or wrong way to do this, I just find the extra LoC and longer doc pages distracting.  Feel free to take it or leave it.

Reply:
> E.g. `DeleteAll` isn't necessary if the resource ID in `Delete` is optional

That seems super-risky. This is how Google managed to reformat a whole DC by having an empty machine ID field mean all machines: https://sre.google/workbook/postmortem-culture/#case-study

I'm not sure how severe a delete-all would be in this case, but I'd still advise against such an API.

Reply:
A delete all of the okta assignments could lead to an unintended privilege escalation in Okta for longer than intended, so it definitely has some risk to fold the deletes together.

I tend to be a bit more verbose with my interfaces for better or for worse, so I'll probably leave them as is. I appreciate the feedback, however!